Taskfile to deploy Infinispan

pruivo opened the pull request:
* Inital commit

Closes #450

ryanemerson commented on provision/infinispan/ispn-helm/values.yaml:
```diff
@@ -0,0 +1,33 @@
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+replicas: 3
```
In the bash scripts this is 2.

pruivo replied:
I don't have a strong opinion if it should be 2 or 3 :shrug:

ryanemerson commented on provision/infinispan/ispn-helm/templates/infinispan.yaml:
```diff
@@ -0,0 +1,92 @@
+  namespace: {{ $namespace }}
+  annotations:
+    infinispan.org/monitoring: 'true'
+spec:
```
We can reduce the footprint of the clusters by disabling the ConfigListener. As we're defining all caches statically, and users can't create new caches at runtime, there's no benefit of enabling this.

```yaml
spec:
  configListener:
    enabled: false
```

pruivo replied:
Added!

ryanemerson commented on provision/infinispan/ispn-helm/templates/infinispan.yaml:
```diff
@@ -0,0 +1,92 @@
+  annotations:
+    infinispan.org/monitoring: 'true'
+spec:
+  affinity:
```
I believe these are the default values right? We can omit these and they will automatically be populated.

ryanemerson commented on provision/infinispan/README.adoc:
```diff
@@ -53,6 +53,28 @@ CLUSTER_1="gh-pruivo" CLUSTER_2="gh-keycloak" NS_1="ispn-server" NS_2="ispn-serv
 
+== Taskfile
+
+For single cluster mode:
```
We should add corresponding delete tasks for each of these tasks.

kami619 replied:
+1 I am not currently sure, what is the best way to delete the ISPN clustered configuration., would I just delete the `Infinispan` namespace(s) and reset the config in the Keycloak deployment `.env` ?

pruivo replied:
Added a task to delete the Infinispan CR. Didn't delete the namespace because I've been running keycloak and infinispan in the same namespace.

ryanemerson commented on provision/infinispan/README.adoc:
```diff
@@ -53,6 +53,28 @@ CLUSTER_1="gh-pruivo" CLUSTER_2="gh-keycloak" NS_1="ispn-server" NS_2="ispn-serv
+And finally, for cross-site between different OCP clusters:
+[source, bash]
+----
+task crossdc-single ROSA_CLUSTER_NAME_1=gh-keycloak ROSA_CLUSTER_NAME_2=gh-pruivo OC_NAMESPACE_1=infinispan OC_NAMESPACE_2=infinispan
```
`task crossdc ROSA_CLUSTER_NAME_1=gh-keycloak ...`

pruivo replied:
right :rofl:

ryanemerson reviewed (commented):
The 3 taskfile commands are working as expected :tada:

I think the main tasks left are:

1. Add cleanup tasks to the taskfile.
2. Remove redundant bash scripts and docs.

kami619 commented:
https://github.com/keycloak/keycloak-benchmark/blob/6bb929052500ab5f3b54ea12f537f9d4ec391d59/provision/infinispan/README.adoc?plain=1#L108

do we want to change this to a task command to keep things consistent ? @pruivo

kami619 commented:
```$ task delete-infinispan ROSA_CLUSTER_NAME=gh-keycloak OC_NAMESPACE=infinispan
task: [rosa-oc-login] mkdir -p .task/kubecfg
task: [rosa-oc-login] KUBECONFIG=".task/kubecfg/delete" CLUSTER_NAME="gh-keycloak" ../aws/rosa_oc_login.sh
[rosa-oc-login] W0802 10:54:30.001835  465828 loader.go:222] Config not found: .task/kubecfg/delete
[rosa-oc-login] WARNING: Using insecure TLS client config. Setting this option is not supported!
[rosa-oc-login] 
[rosa-oc-login] Login successful.
[rosa-oc-login] 
[rosa-oc-login] You have access to 105 projects, the list has been suppressed. You can list all projects with 'oc projects'
[rosa-oc-login] 
[rosa-oc-login] Using project "default".
[rosa-oc-login] W0802 10:54:32.110782  465828 loader.go:222] Config not found: .task/kubecfg/delete
[rosa-oc-login] W0802 10:54:32.110905  465828 loader.go:222] Config not found: .task/kubecfg/delete
task: [delete-infinispan] KUBECONFIG=".task/kubecfg/delete" oc delete infinispans.infinispan.org -n "infinispan" infinispan
[delete-infinispan] infinispan.infinispan.org "infinispan" deleted

$ task single-cluster ROSA_CLUSTER_NAME=gh-keycloak OC_NAMESPACE=infinispan
task: Task "single-cluster" is up to date
```

@pruivo once we delete the infinispan CR and then try to recreate it, the dirty check in the Task files seems to be miss to track the changes done to deleting the namespace and it thinks the last run task had no change. should we track this somehow ?

kami619 commented on provision/infinispan/README.adoc:
```diff
@@ -2,55 +2,26 @@
 [source, bash]
 ----
-CLUSTER_1="gh-pruivo" CLUSTER_2="gh-keycloak" NS_1="ispn-server" NS_2="ispn-server" ./create_ispn_clusters.sh
+task crossdc-single ROSA_CLUSTER_NAME=gh-keycloak OC_NAMESPACE_1=ispn-1 OC_NAMESPACE_2=ispn-2
```
can we re-use the existing infinispan namespaces to run this command ?

pruivo replied:
I've been using the same namespace for keycloak and infinispan without any issues.

kami619 commented:
```
$ task -f delete-infinispan ROSA_CLUSTER_NAME=gh-keycloak OC_NAMESPACE=infinispan
task: [rosa-oc-login] mkdir -p .task/kubecfg
task: [rosa-oc-login] KUBECONFIG=".task/kubecfg/delete" CLUSTER_NAME="gh-keycloak" ../aws/rosa_oc_login.sh
[rosa-oc-login] WARNING: Using insecure TLS client config. Setting this option is not supported!
[rosa-oc-login] 
[rosa-oc-login] Login successful.
[rosa-oc-login] 
[rosa-oc-login] You have access to 106 projects, the list has been suppressed. You can list all projects with 'oc projects'
[rosa-oc-login] 
[rosa-oc-login] Using project "default".
task: [delete-infinispan] KUBECONFIG=".task/kubecfg/delete" oc delete infinispans.infinispan.org -n "infinispan" infinispan
[delete-infinispan] infinispan.infinispan.org "infinispan" deleted

```

When I run this `delete-infinispan` task and I see the log message that its deleted, I still see the namespace to be existing in the oc projects list, are we missing a clean-up step or should we wait long enough for the namespace to be fully deleted ?

```
$ oc projects | grep infinispan
    infinispan

```

kami619 commented on provision/infinispan/README.adoc:
```diff
@@ -2,55 +2,26 @@
 [source, bash]
 ----
-CLUSTER_1="gh-pruivo" CLUSTER_2="gh-keycloak" NS_1="server-site-1" NS_2="server-site-2" ./create_ispn_clusters.sh
+task crossdc ROSA_CLUSTER_NAME_1=gh-keycloak ROSA_CLUSTER_NAME_2=gh-pruivo OC_NAMESPACE_1=infinispan OC_NAMESPACE_2=infinispan
```
can we add some example log message that would reflect a successful creation of this cross-site in all three situations of.,

- single-cluster
- crossdc-single
- crossdc

pruivo replied:
added waiting conditions. It should print the following
```
task: [wait-cluster] echo "Waiting for the Infinispan cluster to form on namespace 'pruivo1' (timeout after 300 seconds)"
[wait-cluster] Waiting for the Infinispan cluster to form on namespace 'pruivo1' (timeout after 300 seconds)
task: [wait-cluster] KUBECONFIG=".task/kubecfg/crossdcsingle" oc wait --for condition=WellFormed --timeout=300s infinispans.infinispan.org -n "pruivo1" infinispan
[wait-cluster] infinispan.infinispan.org/infinispan condition met
task: [wait-cluster] echo "Waiting for the Infinispan cluster to form on namespace 'pruivo2' (timeout after 300 seconds)"
[wait-cluster] Waiting for the Infinispan cluster to form on namespace 'pruivo2' (timeout after 300 seconds)
task: [wait-cluster] KUBECONFIG=".task/kubecfg/crossdcsingle" oc wait --for condition=WellFormed --timeout=300s infinispans.infinispan.org -n "pruivo2" infinispan
[wait-cluster] infinispan.infinispan.org/infinispan condition met
task: [wait-crossdc] echo "Waiting for the Infinispan cross-site to form on namespace 'pruivo1' (timeout after 300 seconds)"
[wait-crossdc] Waiting for the Infinispan cross-site to form on namespace 'pruivo1' (timeout after 300 seconds)
task: [wait-crossdc] KUBECONFIG=".task/kubecfg/crossdcsingle" oc wait --for condition=CrossSiteViewFormed --timeout=300s infinispans.infinispan.org -n "pruivo1" infinispan
[wait-crossdc] infinispan.infinispan.org/infinispan condition met
task: [wait-crossdc] echo "Waiting for the Infinispan cross-site to form on namespace 'pruivo2' (timeout after 300 seconds)"
[wait-crossdc] Waiting for the Infinispan cross-site to form on namespace 'pruivo2' (timeout after 300 seconds)
task: [wait-crossdc] KUBECONFIG=".task/kubecfg/crossdcsingle" oc wait --for condition=CrossSiteViewFormed --timeout=300s infinispans.infinispan.org -n "pruivo2" infinispan
[wait-crossdc] infinispan.infinispan.org/infinispan condition met
```

kami619 replied:
Thanks for adding the wait and listing out the actual output from the task output and at the sametime, any hint from the Keycloak server on how the logs look for a successful cross-site creation @pruivo

pruivo commented:
@kami619 the namespace is not removed since it can be shared. I'm having a hard time understanding the semantics of `source` and `status`. Also, I'm not sure what `common:split` is supposed to do but since it is used everywhere, I added it here too.

mhajas commented:
@pruivo All of these settings serve to say when to rerun the task and when it is not necessary:
- **`source`** - says these files are a source for this task, if the files change between two runs this task is executed again, otherwise, it is not
- **`status`** - when the condition is false the task executed otherwise it is not
- **`common:split`** - this is `kcb` specific mechanism to detect changes in tasks within each Taskfile definition. When you make some change in some task, it is executed again together with all tasks that depend on it. The `common:split` task takes the Taskfile and cut it to one file per task, and then each of these files is set as a source for the corresponding task using: 
```
    sources:
      - .task/subtask-{{.TASK}}.yaml
```

kami619 commented on provision/infinispan/README.adoc:
```diff
@@ -2,55 +2,26 @@
 [source, bash]
 ----
-CLUSTER_1="gh-pruivo" CLUSTER_2="gh-keycloak" NS_1="ispn-server" NS_2="ispn-server" ./create_ispn_clusters.sh
+task crossdc-single ROSA_CLUSTER_NAME=gh-keycloak OC_NAMESPACE_1=ispn-1 OC_NAMESPACE_2=ispn-2
```
I brought up the ISPN cluster with 

`task crossdc-single ROSA_CLUSTER_NAME=gh-keycloak OC_NAMESPACE_1=ispn-1 OC_NAMESPACE_2=ispn-2` 

@pruivo Once we deployed Keycloak with below configuration, is it possible to switch to a different ISPN namespace in Keycloak i.e. change it to `KC_ISPN_NAMESPACE=ispn-2` and re-deploy. And currently if I do that, I end up in an error state for one of the Keycloak Pods.

```
#ISPN
KC_CUSTOM_INFINISPAN_CONFIG_FILE=config/kcb-infinispan-cache-remote-store-config.xml
KC_ISPN_NAMESPACE=ispn-1
KC_ISPN_CLUSTER=infinispan

#DB
KC_DATABASE=postgres
KC_DB_POOL_MIN_SIZE=15
KC_DB_POOL_MAX_SIZE=15
KC_DB_POOL_INITIAL_SIZE=15
KC_STORAGE=

```

kami619 replied:
```
"message":"Error details:","threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-2","processName":"QuarkusEntryPoint","processId":1,"exception":{"refId":1,"exceptionType":"java.lang.RuntimeException","message":"Failed to start caches","frames":

[{"class":"org.keycloak.quarkus.runtime.storage.legacy.infinispan.CacheManagerFactory","method":"getOrCreate","line":61},{"class":"org.keycloak.quarkus.runtime.storage.legacy.infinispan.CacheManagerFactory_9497dc12cf1c9738de11a9c4b512d17ae927dceb_Synthetic_ClientProxy","method":"getOrCreate"}
```

kami619 replied:
yup I did use --force flag

pruivo replied:
isn't the cause available in the logs?

pruivo commented:
Split the `Taskfile.yaml` in two different yaml files to ease in the includes.
`Taskfile.yaml` contains only the public tasks to be invoked:
```
task: Available tasks for this project:
* crossdc:                     Creates a cross-site enabled Infinispan deployment between 2 OCP clusters
* crossdc-single:              Creates a cross-site enabled Infinispan deployment between 2 namespaces
* delete-crossdc:              Deletes the Infinispan CR from cross-site deployment
* delete-crossdc-single:       Deletes the Infinispan CR in a cross-site deployment in a single OCP cluster
* delete-infinispan:           Deletes the Infinispan CR
* single-cluster:              Creates a single cluster Infinispan deployment
* wait-cluster:                Waits for the Infinispan cluster to form
* wait-crossdc:                Waits for the Infinispan Cross-Site to form
```
`Utils.yaml` contains all the subtasks and can be included in other `Taskfiles.yaml`.
